File: python/해시/완주하지 못한 선수.py

""" 레벨 1. 완주하지 못한 선수 """

def solution(participant, completion):

    answer = {}

    for p in participant:
        answer[p] = answer.setdefault(p, 0) + 1
    
    for c in completion:
        answer[c] -= 1

    for name, count in answer.items():
        if count == 1:
            return name

print(solution(["leo", "kiki", "eden"], ["eden", "kiki"]))                                                              # "leo"
print(solution(["marina", "josipa", "nikola", "vinko", "filipa"], 	["josipa", "filipa", "marina", "nikola"]))          # "vinko" 
print(solution(["mislav", "stanko", "mislav", "ana"], 	["stanko", "ana", "mislav"]))                                   # "mislav"

File: Java/src/자료구조/완주하지못한선수.java
package 자료구조;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

public class 완주하지못한선수 {

    public static void main(String[] args) {

        String[] p1 = {"leo", "kiki", "eden"};
        String[] c1 = {"eden", "kiki"};
        System.out.println(solution(p1, c1));

        String[] p2 = {"marina", "josipa", "nikola", "vinko", "filipa"};
        String[] c2 = {"josipa", "filipa", "marina", "nikola"};
        System.out.println(solution(p2, c2));

        String[] p3 = {"mislav", "stanko", "mislav", "ana"};
        String[] c3 = {"stanko", "ana", "mislav"};
        System.out.println(solution(p3, c3));

    }

    public static String solution(String[] participant, String[] completion) {

        HashMap<String, Integer> map = new HashMap<>();
        
        for (String p : participant) {
            map.put(p, map.getOrDefault(p, 0) + 1);
        }

        for (String c : completion) {
            map.replace(c, map.get(c) - 1);
        }

        Iterator<Map.Entry<String, Integer>> iter = map.entrySet().iterator();
        while (iter.hasNext()) {
            Map.Entry<String, Integer> entry = iter.next();
            if (entry.getValue() > 0) {
                return entry.getKey().toString();
            }
        }

        return null;
    }
}
